cc/gtp_player.cc

```diff
@@ -0,0 +1,324 @@
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  return Response::Ok(absl::StrJoin(args, " "));
+}
+
+GtpPlayer::Response GtpPlayer::HandleFinalScore(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  auto response = CheckArgsExact(cmd, 0, args);
+  if (!response.ok) {
+    return response;
+  }
+  if (!game_over()) {
+    return Response::Error("game isn't over");
+  } else {
+    return Response::Ok(result_string());
+  }
+}
+
+GtpPlayer::Response GtpPlayer::HandleGamestate(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  auto response = CheckArgsExact(cmd, 0, args);
+  if (!response.ok) {
+    return response;
+  }
+
+  const auto& position = root()->position;
+
+  // board field.
+  std::ostringstream oss;
+  for (const auto& stone : position.stones()) {
+    char ch;
+    if (stone.color() == Color::kBlack) {
+      ch = 'X';
+    } else if (stone.color() == Color::kWhite) {
+      ch = 'O';
+    } else {
+      ch = '.';
+    }
+    oss << ch;
+  }
+  std::string board = oss.str();
+
+  // toPlay field.
+  std::string to_play = position.to_play() == Color::kBlack ? "Black" : "White";
+
+  // lastMove field.
+  std::string last_move;
+  if (!history().empty()) {
+    last_move = absl::StrCat("\"", history().back().c.ToKgs(), "\"");
+  } else {
+    last_move = "null";
+  }
+
+  // n field.
+  auto n = position.n();
+
+  // q field.
+  auto q = root()->parent != nullptr ? root()->parent->Q() : 0;
+
+  // clang-format off
+  std::cerr << "mg-gamestate: {"
+            << "\"board\":\"" << board << "\", "
+            << "\"toPlay\":\"" << to_play << "\", "
+            << "\"lastMove\":" << last_move << ", "
+            << "\"n\":" << n << ", "
+            << "\"q\":" << q
+            << "}" << std::endl;
+  // clang-format on
+  return Response::Ok();
+}
+
+GtpPlayer::Response GtpPlayer::HandleGenmove(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  auto response = CheckArgsRange(cmd, 0, 1, args);
+  if (!response.ok) {
+    return response;
+  }
+
+  auto c = SuggestMove(num_readouts_);
+  std::cerr << root()->Describe() << std::endl;
+  PlayMove(c);
+
+  return Response::Ok(c.ToKgs());
+}
+
+GtpPlayer::Response GtpPlayer::HandleName(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  auto response = CheckArgsExact(cmd, 0, args);
+  if (!response.ok) {
+    return response;
+  }
+  return Response::Ok(name_);
+}
+
+GtpPlayer::Response GtpPlayer::HandlePlay(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  auto response = CheckArgsExact(cmd, 2, args);
+  if (!response.ok) {
+    return response;
+  }
+
+  Color color;
+  if (std::tolower(args[0][0]) == 'b') {
+    color = Color::kBlack;
+  } else if (std::tolower(args[0][0]) == 'w') {
+    color = Color::kWhite;
+  } else {
+    return Response::Error("expected b or w for player color, got ", args[0]);
```

Comment:
You might think it is handy to be able to fail with useful, descriptive error messages.  but alas, the spec constrains the error messages.  For these other cases, printing something actually useful on stderr and failing on stdout with "illegal move" is correct behavior.

Reply:
done